File: wpilibcIntegrationTests/src/main/native/cpp/shuffleboard/ShuffleboardTabTest.cpp
/*----------------------------------------------------------------------------*/
/* Copyright (c) 2018 FIRST. All Rights Reserved.                             */
/* Open Source Software - may be modified and shared by FRC teams. The code   */
/* must be accompanied by the FIRST BSD license file in the root directory of */
/* the project.                                                               */
/*----------------------------------------------------------------------------*/

#include <array>
#include <memory>
#include <string>

#include <networktables/NetworkTableEntry.h>
#include <networktables/NetworkTableInstance.h>

#include "frc/commands/InstantCommand.h"
#include "frc/shuffleboard/ShuffleboardInstance.h"
#include "frc/shuffleboard/ShuffleboardTab.h"
#include "frc/smartdashboard/Sendable.h"
#include "gtest/gtest.h"

using namespace frc;

class ShuffleboardTabTest : public testing::Test {
  void SetUp() override {
    m_ntInstance = nt::NetworkTableInstance::Create();
    m_instance = std::make_unique<detail::ShuffleboardInstance>(m_ntInstance);
    m_tab = &(m_instance->GetTab("Tab"));
  }

 protected:
  nt::NetworkTableInstance m_ntInstance;
  ShuffleboardTab* m_tab;
  std::unique_ptr<detail::ShuffleboardInstance> m_instance;
};

TEST_F(ShuffleboardTabTest, AddDouble) {
  auto entry = m_tab->Add("Double", 1.0).GetEntry();
  EXPECT_EQ("/Shuffleboard/Tab/Double", entry.GetName());
  EXPECT_FLOAT_EQ(1.0, entry.GetValue()->GetDouble());
}

TEST_F(ShuffleboardTabTest, AddInteger) {
  auto entry = m_tab->Add("Int", 1).GetEntry();
  EXPECT_EQ("/Shuffleboard/Tab/Int", entry.GetName());
  EXPECT_FLOAT_EQ(1.0, entry.GetValue()->GetDouble());
}

TEST_F(ShuffleboardTabTest, AddBoolean) {
  auto entry = m_tab->Add("Bool", false).GetEntry();
  EXPECT_EQ("/Shuffleboard/Tab/Bool", entry.GetName());
  EXPECT_FALSE(entry.GetValue()->GetBoolean());
}

TEST_F(ShuffleboardTabTest, AddString) {
  auto entry = m_tab->Add("String", "foobar").GetEntry();
  EXPECT_EQ("/Shuffleboard/Tab/String", entry.GetName());
  EXPECT_EQ("foobar", entry.GetValue()->GetString());
}

TEST_F(ShuffleboardTabTest, AddNamedSendableWithProperties) {
  InstantCommand sendable("Command");
  std::string widgetType = "Command Widget";
  wpi::StringMap<std::shared_ptr<nt::Value>> map;
  map.try_emplace("foo", nt::Value::MakeDouble(1234));
  map.try_emplace("bar", nt::Value::MakeString("baz"));
  m_tab->Add(sendable).WithWidget(widgetType).WithProperties(map);

  m_instance->Update();
  std::string meta = "/Shuffleboard/.metadata/Tab/Command";

  EXPECT_EQ(1234, m_ntInstance.GetEntry(meta + "/Properties/foo").GetDouble(-1))
      << "Property 'foo' not set correctly";
  EXPECT_EQ("baz",
            m_ntInstance.GetEntry(meta + "/Properties/bar").GetString(""))
      << "Property 'bar' not set correctly";
  EXPECT_EQ(widgetType,
            m_ntInstance.GetEntry(meta + "/PreferredComponent").GetString(""))
      << "Preferred component not set correctly";
}

TEST_F(ShuffleboardTabTest, AddNumberArray) {
  std::array<double, 3> expect = {{1.0, 2.0, 3.0}};
  auto entry = m_tab->Add("DoubleArray", expect).GetEntry();
  EXPECT_EQ("/Shuffleboard/Tab/DoubleArray", entry.GetName());

  auto actual = entry.GetValue()->GetDoubleArray();
  EXPECT_EQ(expect.size(), actual.size());
  for (size_t i = 0; i < expect.size(); i++) {
    EXPECT_FLOAT_EQ(expect[i], actual[i]);
  }
}

TEST_F(ShuffleboardTabTest, AddBooleanArray) {
  std::array<bool, 2> expect = {{true, false}};
  auto entry = m_tab->Add("BoolArray", expect).GetEntry();
  EXPECT_EQ("/Shuffleboard/Tab/BoolArray", entry.GetName());

  auto actual = entry.GetValue()->GetBooleanArray();
  EXPECT_EQ(expect.size(), actual.size());
  for (size_t i = 0; i < expect.size(); i++) {
    EXPECT_EQ(expect[i], actual[i]);
  }
}

TEST_F(ShuffleboardTabTest, AddStringArray) {
  std::array<std::string, 2> expect = {{"foo", "bar"}};
  auto entry = m_tab->Add("StringArray", expect).GetEntry();
  EXPECT_EQ("/Shuffleboard/Tab/StringArray", entry.GetName());

  auto actual = entry.GetValue()->GetStringArray();
  EXPECT_EQ(expect.size(), actual.size());
  for (size_t i = 0; i < expect.size(); i++) {
    EXPECT_EQ(expect[i], actual[i]);
  }
}

File: wpilibOldCommands/src/test/java/edu/wpi/first/wpilibj/shuffleboard/ShuffleboardTabTest.java
/*----------------------------------------------------------------------------*/
/* Copyright (c) 2018-2020 FIRST. All Rights Reserved.                        */
/* Open Source Software - may be modified and shared by FRC teams. The code   */
/* must be accompanied by the FIRST BSD license file in the root directory of */
/* the project.                                                               */
/*----------------------------------------------------------------------------*/

package edu.wpi.first.wpilibj.shuffleboard;

import java.util.HashMap;
import java.util.Map;

import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import edu.wpi.first.networktables.NetworkTableEntry;
import edu.wpi.first.networktables.NetworkTableInstance;
import edu.wpi.first.wpilibj.Sendable;
import edu.wpi.first.wpilibj.command.InstantCommand;

import static org.junit.jupiter.api.Assertions.assertAll;
import static org.junit.jupiter.api.Assertions.assertArrayEquals;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;

@SuppressWarnings({"PMD.TooManyMethods"})
public class ShuffleboardTabTest {
  private NetworkTableInstance m_ntInstance;
  private ShuffleboardTab m_tab;
  private ShuffleboardInstance m_instance;

  @BeforeEach
  void setup() {
    m_ntInstance = NetworkTableInstance.create();
    m_instance = new ShuffleboardInstance(m_ntInstance);
    m_tab = m_instance.getTab("Tab");
  }

  @AfterEach
  void tearDown() {
    m_ntInstance.close();
  }

  @Test
  void testAddDouble() {
    NetworkTableEntry entry = m_tab.add("Double", 1.0).getEntry();
    assertAll(
        () -> assertEquals("/Shuffleboard/Tab/Double", entry.getName()),
        () -> assertEquals(1.0, entry.getValue().getDouble()));
  }

  @Test
  void testAddInteger() {
    NetworkTableEntry entry = m_tab.add("Int", 1).getEntry();
    assertAll(
        () -> assertEquals("/Shuffleboard/Tab/Int", entry.getName()),
        () -> assertEquals(1.0, entry.getValue().getDouble()));
  }

  @Test
  void testAddLong() {
    NetworkTableEntry entry = m_tab.add("Long", 1L).getEntry();
    assertAll(
        () -> assertEquals("/Shuffleboard/Tab/Long", entry.getName()),
        () -> assertEquals(1.0, entry.getValue().getDouble()));
  }


  @Test
  void testAddBoolean() {
    NetworkTableEntry entry = m_tab.add("Bool", false).getEntry();
    assertAll(
        () -> assertEquals("/Shuffleboard/Tab/Bool", entry.getName()),
        () -> assertFalse(entry.getValue().getBoolean()));
  }

  @Test
  void testAddString() {
    NetworkTableEntry entry = m_tab.add("String", "foobar").getEntry();
    assertAll(
        () -> assertEquals("/Shuffleboard/Tab/String", entry.getName()),
        () -> assertEquals("foobar", entry.getValue().getString()));
  }

  @Test
  void testAddNamedSendableWithProperties() {
    Sendable sendable = new InstantCommand("Command");
    String widgetType = "Command Widget";
    m_tab.add(sendable)
       .withWidget(widgetType)
       .withProperties(mapOf("foo", 1234, "bar", "baz"));

    m_instance.update();
    String meta = "/Shuffleboard/.metadata/Tab/Command";

    assertAll(
        () -> assertEquals(1234,
                           m_ntInstance.getEntry(meta + "/Properties/foo").getDouble(-1),
                           "Property 'foo' not set correctly"),
        () -> assertEquals("baz",
                           m_ntInstance.getEntry(meta + "/Properties/bar").getString(null),
                           "Property 'bar' not set correctly"),
        () -> assertEquals(widgetType,
                           m_ntInstance.getEntry(meta + "/PreferredComponent").getString(null),
                           "Preferred component not set correctly"));
  }

  @Test
  void testAddNumberArray() {
    NetworkTableEntry entry = m_tab.add("DoubleArray", new double[]{1, 2, 3}).getEntry();
    assertAll(
        () -> assertEquals("/Shuffleboard/Tab/DoubleArray", entry.getName()),
        () -> assertArrayEquals(new double[]{1, 2, 3}, entry.getValue().getDoubleArray()));
  }

  @Test
  void testAddBooleanArray() {
    NetworkTableEntry entry = m_tab.add("BoolArray", new boolean[]{true, false}).getEntry();
    assertAll(
        () -> assertEquals("/Shuffleboard/Tab/BoolArray", entry.getName()),
        () -> assertArrayEquals(new boolean[]{true, false}, entry.getValue().getBooleanArray()));
  }

  @Test
  void testAddStringArray() {
    NetworkTableEntry entry = m_tab.add("StringArray", new String[]{"foo", "bar"}).getEntry();
    assertAll(
        () -> assertEquals("/Shuffleboard/Tab/StringArray", entry.getName()),
        () -> assertArrayEquals(new String[]{"foo", "bar"}, entry.getValue().getStringArray()));
  }

  /**
   * Stub for Java 9 {@code Map.of()}.
   */
  @SuppressWarnings({"unchecked", "PMD"})
  private static <K, V> Map<K, V> mapOf(Object... entries) {
    Map<K, V> map = new HashMap<>();
    for (int i = 0; i < entries.length; i += 2) {
      map.put((K) entries[i], (V) entries[i + 1]);
    }
    return map;
  }

}
